apps/shared/avifpng.c

```diff
@@ -104,61 +108,132 @@ static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, a
+        if (!*ignoreExif && !strcmp(text->key, "Raw profile type exif")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            uint32_t exifTiffHeaderOffset;
+            if (avifExtractExifTiffHeaderOffset(metadata.data, metadata.size, &exifTiffHeaderOffset) != AVIF_RESULT_OK) {
+                avifRWDataFree(&metadata);
+                fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
+                return AVIF_FALSE;
+            }
+            avifRWDataFree(&avif->exif);
+            avif->exif = metadata;
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+        } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            if (tagExif00WasRemoved) {
+                avifRWDataFree(&metadata);
+                fprintf(stderr, "XMP extraction failed: XMP payload begins with an Exif tag\n");
+                return AVIF_FALSE;
             }
-            return avifCopyRawProfile(text->text, textLength, &avif->exif);
+            avifRWDataFree(&avif->xmp);
+            avif->xmp = metadata;
+            *ignoreXMP = AVIF_TRUE; // Ignore any other XMP chunk.
+        } else if (!strcmp(text->key, "Raw profile type APP1")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            // Differentiate Exif and XMP by checking whether the chunk begins with a valid TIFF header ("IIx0x42" or "MMx42x0").
+            const uint32_t tiffHeaderSize = 4;
+            uint32_t exifTiffHeaderOffset;
+            if (metadata.size > tiffHeaderSize &&
+                (avifExtractExifTiffHeaderOffset(metadata.data, tiffHeaderSize, &exifTiffHeaderOffset) == AVIF_RESULT_OK)) {
+                if (*ignoreExif) {
+                    avifRWDataFree(&metadata);
+                } else {
+                    avifRWDataFree(&avif->exif);
+                    avif->exif = metadata;
+                    *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+                }
+            } else if (tagExif00WasRemoved) {
+                if (*ignoreExif) {
+                    // Ignore invalid Exif payloads too.
+                    avifRWDataFree(&metadata);
+                } else {
+                    avifRWDataFree(&metadata);
+                    fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
+                    return AVIF_FALSE;
+                }
+            } else {
+                // For simplicity, consider everything else as XMP.
+                // In theory, it could be an ICC stored as text but iCCP has been around for a while,
```

Comment:
Don't think it applies here - we're inside the "Raw profile type APP1" case which was only ever used for Exif or XMP it would seem. I'd move this ICC comment to the end of this big decision tree and mention no support for "Raw profile type icc" or "Raw profile type icm" keys.

Reply:
Sorry I mixed them up. I moved the comment near `png_get_iCCP()`.